src/metabase/driver/sql/query_processor.clj

```diff
@@ -1409,23 +1411,44 @@
 (declare apply-clauses)
 
 (defn- apply-source-query
-  "Handle a `:source-query` clause by adding a recursive `SELECT` or native query. At the time of this writing, all
-  source queries are aliased as `source`."
-  [driver honeysql-form {{:keys [native params],
-                          persisted :persisted-info/native
-                          :as source-query} :source-query}]
-  (assoc honeysql-form
-         :from [[(cond
-                   persisted
-                   (sql-source-query persisted nil)
-
-                   native
-                   (sql-source-query native params)
-
-                   :else
-                   (apply-clauses driver {} source-query))
-                 (let [table-alias (->honeysql driver (h2x/identifier :table-alias source-query-alias))]
-                   [table-alias])]]))
+  "Handle a `:source-query` clause by adding a recursive `SELECT` or native query.
+   If the source query has ambiguous column names, use a `WITH` statement to rename the source columns.
+   At the time of this writing, all source queries are aliased as `source`."
+  [driver honeysql-form {{:keys [native params] persisted :persisted-info/native :as source-query} :source-query
+                         source-metadata :source-metadata :as _x}]
+  (let [table-alias (->honeysql driver (h2x/identifier :table-alias source-query-alias))
+        source-clause (cond
+                        persisted
+                        (sql-source-query persisted nil)
+
+                        native
+                        (sql-source-query native params)
+
+                        :else
+                        (apply-clauses driver {} source-query))
+        ;; TODO: Use MLv2 here to get source and desired-aliases
+        alias-info (mapv (fn [{[_ desired-ref-name] :field_ref source-name :name}]
+                           [source-name desired-ref-name])
+                         source-metadata)
+        source-aliases (mapv first alias-info)
+        desired-aliases (mapv second alias-info)
+        duplicate-source-aliases? (not (apply distinct? source-aliases))
```

Comment:
When I run the code with fresh app db, I get an sync error. Further examining what's going on (original exception is not preserved), I discovered the exception is thrown on this line:

```
clojure.lang.ExceptionInfo: Error compiling query: Wrong number of args (0) passed to: clojure.core/distinct? {:query {:database 1, :type :query, :query {:source-table 1, :expressions {"substring172365" [:substring [:field 17 nil] 1 1234], "substring172366" [:substring [:field 18 nil] 1 1234], "substring172367" [:substring [:field 19 nil] 1 1234]}, :fields [[:expression "substring172365"] [:expression "substring172366"] [:expression "substring172367"] [:field 22 {:temporal-unit :default}] [:field 21 nil]], :limit 10000}, :middleware {:format-rows? false, :skip-results-metadata? true}}, :type :driver, :e #error {
 :cause "Wrong number of args (0) passed to: clojure.core/distinct?"
 :via
 [{:type clojure.lang.ArityException
   :message "Wrong number of args (0) passed to: clojure.core/distinct?"
   :at [clojure.lang.AFn throwArity "AFn.java" 429]}]
 :trace
 [[clojure.lang.AFn throwArity "AFn.java" 429]
  [clojure.lang.RestFn invoke "RestFn.java" 399]
  [clojure.lang.AFn applyToHelper "AFn.java" 152]
  [clojure.lang.RestFn applyTo "RestFn.java" 132]
  [clojure.core$apply invokeStatic "core.clj" 667]
  [clojure.core$apply invoke "core.clj" 662]
  [metabase.driver.sql.query_processor$apply_source_query invokeStatic "query_processor.clj" 1435]
  [metabase.driver.sql.query_processor$apply_source_query invoke "query_processor.clj" 1413]
  [metabase.driver.sql.query_processor$apply_clauses invokeStatic "query_processor.clj" 1461]
  ...
```

Problem seems to be, that under specific circumstances we do `(apply distinct? [])` which results in call to `distinct?` with no arguments.

Ids from query to reproduce on (first line of the trace) follows:
```
metabase=# select id, name from metabase_database where id = 1;
 id |      name       
----+-----------------
  1 | Sample Database
(1 row)

metabase=# select id, name from metabase_table where id = 1;                                  
 id |      name       
----+-----------------
  1 | ANALYTIC_EVENTS
(1 row)    

metabase=# select id, name from metabase_field where id > 16 and id < 23;
 id |     name     
----+--------------
 17 | EVENT
 18 | PAGE_URL
 19 | BUTTON_LABEL
 20 | ID
 21 | ACCOUNT_ID
 22 | TIMESTAMP
(6 rows)
```